# Mini Proyecto - Analisis de Datos con Python

Este notebook inicia el analisis tecnico del dataset de ferreteria usado en el Proyecto 1: **Sistema de Business Intelligence - Ferreteria**.

En esta primera version se realiza una lectura exploratoria inicial para confirmar que los archivos se cargan correctamente, revisar estructura, tipos de datos, valores nulos y duplicados.


## 1. Contexto del proyecto

El Proyecto 1 se enfoco en Business Intelligence con Power BI. Este Proyecto 2 usa el mismo dataset, pero desarrolla una capa de analisis con Python para documentar limpieza, transformacion, validacion y exploracion de datos.

Objetivo de esta etapa: entender el estado inicial de los CSV antes de hacer cualquier transformacion.


## 2. Importacion de librerias


In [11]:
from pathlib import Path

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

sns.set_theme(style="whitegrid")

PROJECT_ROOT = Path.cwd().parent if Path.cwd().name == "notebooks" else Path.cwd()
DATA_RAW = PROJECT_ROOT / "data" / "raw"
DATA_PROCESSED = PROJECT_ROOT / "data" / "processed"
FIGURES = PROJECT_ROOT / "reports" / "figures"

print("Ruta del proyecto:", PROJECT_ROOT)


Ruta del proyecto: c:\Users\EDELSON ORIHUELA\OneDrive\Documentos\PROYECTOS CV\GPT\Mini Proyecto – Analisis de Datos con Python


## 3. Carga de datos


In [12]:
productos = pd.read_csv(DATA_RAW / "productos.csv")
ventas = pd.read_csv(DATA_RAW / "ventas.csv")

print("Productos:", productos.shape)
print("Ventas:", ventas.shape)


Productos: (72, 5)
Ventas: (1800, 6)


## 4. Vista preliminar de los datos


In [13]:
productos.head()


,id_producto,nombre,categoría,costo,stock
0,P001,Martillo de acero 16 oz,Herramientas manuales,18.5,34
1,P002,Alicate universal 8 pulgadas,Herramientas manuales,16.2,42
2,P003,Destornillador plano 6 pulgadas,Herramientas manuales,5.4,85
3,P004,Juego de llaves mixtas 8 piezas,Herramientas manuales,42.0,18
4,P005,Cinta métrica 5 m,Herramientas manuales,8.8,57


In [14]:
ventas.head()


,fecha,producto,categoría,cantidad,precio_unitario,total
0,2025-01-01,Arena fina saco 40 kg,Materiales de construcción,12,9.48,113.76
1,2025-01-01,Cinta teflón 12 mm,Gasfitería,1,2.05,2.05
2,2025-01-01,Llave francesa 10 pulgadas,Herramientas manuales,2,39.88,79.76
3,2025-01-01,Perno hexagonal 1/4 ciento,Tornillería y fijaciones,5,19.94,99.70
4,2025-01-01,Pintura látex blanca 1 gl,Pinturas,1,49.83,49.83


## 5. Estructura general


In [15]:
resumen_estructura = pd.DataFrame({
    "tabla": ["productos", "ventas"],
    "filas": [productos.shape[0], ventas.shape[0]],
    "columnas": [productos.shape[1], ventas.shape[1]],
    "columnas_dataset": [", ".join(productos.columns), ", ".join(ventas.columns)],
})

resumen_estructura


,tabla,filas,columnas,columnas_dataset
0,productos,72,5,"id_producto, nombre, categoría, costo, stock"
1,ventas,1800,6,"fecha, producto, categoría, cantidad, precio_u..."


## 6. Tipos de datos


In [16]:
productos.info()


<class 'pandas.DataFrame'>
RangeIndex: 72 entries, 0 to 71
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   id_producto  72 non-null     str    
 1   nombre       72 non-null     str    
 2   categoría    72 non-null     str    
 3   costo        72 non-null     float64
 4   stock        72 non-null     int64  
dtypes: float64(1), int64(1), str(3)
memory usage: 2.9 KB


In [17]:
ventas.info()


<class 'pandas.DataFrame'>
RangeIndex: 1800 entries, 0 to 1799
Data columns (total 6 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   fecha            1800 non-null   str    
 1   producto         1800 non-null   str    
 2   categoría        1800 non-null   str    
 3   cantidad         1800 non-null   int64  
 4   precio_unitario  1800 non-null   float64
 5   total            1800 non-null   float64
dtypes: float64(2), int64(1), str(3)
memory usage: 84.5 KB


## 7. Revision de valores nulos


In [18]:
productos.isna().sum().to_frame("nulos")


,nulos
id_producto,0
nombre,0
categoría,0
costo,0
stock,0


In [19]:
ventas.isna().sum().to_frame("nulos")


,nulos
fecha,0
producto,0
categoría,0
cantidad,0
precio_unitario,0
total,0


## 8. Revision de duplicados


In [20]:
duplicados = pd.DataFrame({
    "revision": [
        "Productos duplicados por id_producto",
        "Productos duplicados por nombre",
        "Filas duplicadas en ventas",
    ],
    "valor": [
        productos["id_producto"].duplicated().sum(),
        productos["nombre"].duplicated().sum(),
        ventas.duplicated().sum(),
    ],
})

duplicados


,revision,valor
0,Productos duplicados por id_producto,0
1,Productos duplicados por nombre,0
2,Filas duplicadas en ventas,0


## 9. Primeras observaciones

- El dataset de productos contiene **72 registros** y **5 columnas**.
- El dataset de ventas contiene **1,800 registros** y **6 columnas**.
- No se observan valores nulos en esta revision inicial.
- No se observan duplicados por `id_producto` ni por `nombre` en productos.
- No se observan filas duplicadas exactas en ventas.
- La columna `fecha` todavia esta como texto y debera convertirse a fecha en la etapa de limpieza.
- Las columnas numericas principales para analisis seran `cantidad`, `precio_unitario`, `total`, `costo` y `stock`.


## 10. Siguiente paso

En la siguiente etapa se trabajara la limpieza y transformacion:

- normalizar nombres de columnas,
- convertir fechas y campos numericos,
- validar `total = cantidad * precio_unitario`,
- unir ventas con productos,
- crear variables de periodo y margen estimado.


## 11. Limpieza y transformacion de datos

En esta fase se preparan los datos para el analisis:

- normalizacion de nombres de columnas,
- limpieza de textos,
- conversion de fechas y valores numericos,
- validacion del calculo de ventas,
- integracion de ventas con el catalogo de productos,
- creacion de variables derivadas para analisis.


In [21]:
import unicodedata

def normalizar_columna(nombre_columna):
    texto = unicodedata.normalize("NFKD", nombre_columna)
    texto = "".join(caracter for caracter in texto if not unicodedata.combining(caracter))
    texto = texto.strip().lower().replace(" ", "_")
    return texto

productos_limpio = productos.copy()
ventas_limpio = ventas.copy()

productos_limpio.columns = [normalizar_columna(col) for col in productos_limpio.columns]
ventas_limpio.columns = [normalizar_columna(col) for col in ventas_limpio.columns]

print(productos_limpio.columns.tolist())
print(ventas_limpio.columns.tolist())


['id_producto', 'nombre', 'categoria', 'costo', 'stock']
['fecha', 'producto', 'categoria', 'cantidad', 'precio_unitario', 'total']


### 11.1 Limpieza de textos y conversion de tipos

Despues de normalizar los nombres de columnas, se limpian espacios innecesarios en campos de texto y se convierten fechas y columnas numericas al tipo correcto.


In [22]:
# Limpieza de espacios en columnas de texto
for columna in productos_limpio.select_dtypes(include="object").columns:
    productos_limpio[columna] = productos_limpio[columna].astype(str).str.strip()

for columna in ventas_limpio.select_dtypes(include="object").columns:
    ventas_limpio[columna] = ventas_limpio[columna].astype(str).str.strip()

# Conversion de tipos
productos_limpio["costo"] = pd.to_numeric(productos_limpio["costo"], errors="coerce")
productos_limpio["stock"] = pd.to_numeric(productos_limpio["stock"], errors="coerce").astype("Int64")

ventas_limpio["fecha"] = pd.to_datetime(ventas_limpio["fecha"], errors="coerce")
ventas_limpio["cantidad"] = pd.to_numeric(ventas_limpio["cantidad"], errors="coerce").astype("Int64")
ventas_limpio["precio_unitario"] = pd.to_numeric(ventas_limpio["precio_unitario"], errors="coerce")
ventas_limpio["total"] = pd.to_numeric(ventas_limpio["total"], errors="coerce")

print(productos_limpio.dtypes)
print()
print(ventas_limpio.dtypes)


C:\Users\EDELSON ORIHUELA\AppData\Local\Temp\ipykernel_17180\1549277048.py:2: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  for columna in productos_limpio.select_dtypes(include="object").columns:
C:\Users\EDELSON ORIHUELA\AppData\Local\Temp\ipykernel_17180\1549277048.py:5: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See ht

id_producto        str
nombre             str
categoria          str
costo          float64
stock            Int64
dtype: object

fecha              datetime64[us]
producto                      str
categoria                     str
cantidad                    Int64
precio_unitario           float64
total                     float64
dtype: object


### 11.2 Validacion del total de ventas

Se recalcula el total de cada linea de venta para verificar si coincide con el valor informado en el dataset original.


In [23]:
ventas_limpio["total_calculado"] = (
    ventas_limpio["cantidad"].astype(float) * ventas_limpio["precio_unitario"]
).round(2)

ventas_limpio["diferencia_total"] = (
    ventas_limpio["total"] - ventas_limpio["total_calculado"]
).round(2)

errores_total = ventas_limpio[ventas_limpio["diferencia_total"].abs() > 0.01]

print("Registros con diferencia en total:", len(errores_total))
display(ventas_limpio[["cantidad", "precio_unitario", "total", "total_calculado", "diferencia_total"]].head())


Registros con diferencia en total: 0


,cantidad,precio_unitario,total,total_calculado,diferencia_total
0,12,9.48,113.76,113.76,0.0
1,1,2.05,2.05,2.05,0.0
2,2,39.88,79.76,79.76,0.0
3,5,19.94,99.70,99.70,0.0
4,1,49.83,49.83,49.83,0.0


### 11.3 Integracion de ventas con catalogo de productos

Se integra el archivo de ventas con el catalogo de productos para agregar informacion clave como ID de producto, costo unitario y stock actual.


In [24]:
productos_base = productos_limpio.rename(columns={"nombre": "producto"})

ventas_enriquecidas = ventas_limpio.merge(
    productos_base[["id_producto", "producto", "categoria", "costo", "stock"]],
    on="producto",
    how="left",
    suffixes=("_venta", "_catalogo")
)

ventas_sin_producto = ventas_enriquecidas[ventas_enriquecidas["id_producto"].isna()]

print("Ventas sin producto encontrado en catalogo:", len(ventas_sin_producto))
display(ventas_enriquecidas.head())


Ventas sin producto encontrado en catalogo: 0


,fecha,producto,categoria_venta,cantidad,precio_unitario,total,total_calculado,diferencia_total,id_producto,categoria_catalogo,costo,stock
0,2025-01-01,Arena fina saco 40 kg,Materiales de construcción,12,9.48,113.76,113.76,0.0,P018,Materiales de construcción,7.6,120
1,2025-01-01,Cinta teflón 12 mm,Gasfitería,1,2.05,2.05,2.05,0.0,P044,Gasfitería,1.1,210
2,2025-01-01,Llave francesa 10 pulgadas,Herramientas manuales,2,39.88,79.76,79.76,0.0,P008,Herramientas manuales,28.0,16
3,2025-01-01,Perno hexagonal 1/4 ciento,Tornillería y fijaciones,5,19.94,99.70,99.70,0.0,P052,Tornillería y fijaciones,12.5,58
4,2025-01-01,Pintura látex blanca 1 gl,Pinturas,1,49.83,49.83,49.83,0.0,P025,Pinturas,34.0,38


### 11.4 Creacion de variables derivadas

Se crean nuevas variables que facilitan el analisis posterior: margen estimado, margen porcentual y columnas de tiempo para tendencias.


In [25]:
ventas_enriquecidas["categoria"] = ventas_enriquecidas["categoria_catalogo"].fillna(
    ventas_enriquecidas["categoria_venta"]
)

ventas_enriquecidas["margen_unitario"] = (
    ventas_enriquecidas["precio_unitario"] - ventas_enriquecidas["costo"]
).round(2)

ventas_enriquecidas["margen_estimado"] = (
    ventas_enriquecidas["margen_unitario"] * ventas_enriquecidas["cantidad"].astype(float)
).round(2)

ventas_enriquecidas["margen_pct"] = (
    ventas_enriquecidas["margen_estimado"] / ventas_enriquecidas["total"]
).round(4)

ventas_enriquecidas["anio"] = ventas_enriquecidas["fecha"].dt.year
ventas_enriquecidas["mes"] = ventas_enriquecidas["fecha"].dt.month
ventas_enriquecidas["periodo"] = ventas_enriquecidas["fecha"].dt.to_period("M").astype(str)
ventas_enriquecidas["trimestre"] = ventas_enriquecidas["fecha"].dt.to_period("Q").astype(str)

display(
    ventas_enriquecidas[
        [
            "fecha",
            "periodo",
            "producto",
            "categoria",
            "cantidad",
            "precio_unitario",
            "costo",
            "total",
            "margen_unitario",
            "margen_estimado",
            "margen_pct",
        ]
    ].head()
)


,fecha,periodo,producto,categoria,cantidad,precio_unitario,costo,total,margen_unitario,margen_estimado,margen_pct
0,2025-01-01,2025-01,Arena fina saco 40 kg,Materiales de construcción,12,9.48,7.6,113.76,1.88,22.56,0.1983
1,2025-01-01,2025-01,Cinta teflón 12 mm,Gasfitería,1,2.05,1.1,2.05,0.95,0.95,0.4634
2,2025-01-01,2025-01,Llave francesa 10 pulgadas,Herramientas manuales,2,39.88,28.0,79.76,11.88,23.76,0.2979
3,2025-01-01,2025-01,Perno hexagonal 1/4 ciento,Tornillería y fijaciones,5,19.94,12.5,99.70,7.44,37.20,0.3731
4,2025-01-01,2025-01,Pintura látex blanca 1 gl,Pinturas,1,49.83,34.0,49.83,15.83,15.83,0.3177


### 11.5 Generacion del dataset limpio

Se seleccionan y ordenan las columnas finales para crear un dataset limpio, enriquecido y listo para analisis.


In [26]:
ventas_limpias = ventas_enriquecidas[
    [
        "fecha",
        "periodo",
        "anio",
        "mes",
        "trimestre",
        "id_producto",
        "producto",
        "categoria",
        "cantidad",
        "precio_unitario",
        "total",
        "total_calculado",
        "diferencia_total",
        "costo",
        "margen_unitario",
        "margen_estimado",
        "margen_pct",
        "stock",
    ]
].sort_values(["fecha", "producto"]).reset_index(drop=True)

productos_limpios = productos_base.copy()
productos_limpios["valor_inventario"] = (
    productos_limpios["costo"] * productos_limpios["stock"]
).round(2)

ventas_limpias.to_csv(DATA_PROCESSED / "ventas_limpias.csv", index=False, encoding="utf-8-sig")
productos_limpios.to_csv(DATA_PROCESSED / "productos_limpios.csv", index=False, encoding="utf-8-sig")

print("Archivo generado:", DATA_PROCESSED / "ventas_limpias.csv")
print("Archivo generado:", DATA_PROCESSED / "productos_limpios.csv")
print("Ventas limpias:", ventas_limpias.shape)
print("Productos limpios:", productos_limpios.shape)


Archivo generado: c:\Users\EDELSON ORIHUELA\OneDrive\Documentos\PROYECTOS CV\GPT\Mini Proyecto – Analisis de Datos con Python\data\processed\ventas_limpias.csv
Archivo generado: c:\Users\EDELSON ORIHUELA\OneDrive\Documentos\PROYECTOS CV\GPT\Mini Proyecto – Analisis de Datos con Python\data\processed\productos_limpios.csv
Ventas limpias: (1800, 18)
Productos limpios: (72, 6)


### 11.6 Reporte de calidad posterior a la limpieza

Se genera un resumen de validaciones clave para confirmar que el dataset limpio conserva integridad y consistencia.


In [27]:
reporte_calidad = pd.DataFrame(
    [
        {
            "validacion": "productos_total",
            "valor": len(productos_limpios),
            "estado": "OK",
            "detalle": "Cantidad de productos en catalogo limpio",
        },
        {
            "validacion": "ventas_total",
            "valor": len(ventas_limpias),
            "estado": "OK",
            "detalle": "Cantidad de registros en ventas limpias",
        },
        {
            "validacion": "nulos_ventas_limpias",
            "valor": int(ventas_limpias.isna().sum().sum()),
            "estado": "OK" if int(ventas_limpias.isna().sum().sum()) == 0 else "Revisar",
            "detalle": "Valores nulos en dataset limpio de ventas",
        },
        {
            "validacion": "nulos_productos_limpios",
            "valor": int(productos_limpios.isna().sum().sum()),
            "estado": "OK" if int(productos_limpios.isna().sum().sum()) == 0 else "Revisar",
            "detalle": "Valores nulos en dataset limpio de productos",
        },
        {
            "validacion": "ventas_sin_producto",
            "valor": int(ventas_limpias["id_producto"].isna().sum()),
            "estado": "OK" if int(ventas_limpias["id_producto"].isna().sum()) == 0 else "Revisar",
            "detalle": "Ventas sin producto asociado",
        },
        {
            "validacion": "totales_mal_calculados",
            "valor": int((ventas_limpias["diferencia_total"].abs() > 0.01).sum()),
            "estado": "OK" if int((ventas_limpias["diferencia_total"].abs() > 0.01).sum()) == 0 else "Revisar",
            "detalle": "Diferencia entre total original y total recalculado",
        },
    ]
)

display(reporte_calidad)

reporte_calidad.to_csv(PROJECT_ROOT / "reports" / "quality_report.csv", index=False, encoding="utf-8-sig")
print("Reporte generado:", PROJECT_ROOT / "reports" / "quality_report.csv")


,validacion,valor,estado,detalle
0,productos_total,72,OK,Cantidad de productos en catalogo limpio
1,ventas_total,1800,OK,Cantidad de registros en ventas limpias
2,nulos_ventas_limpias,0,OK,Valores nulos en dataset limpio de ventas
3,nulos_productos_limpios,0,OK,Valores nulos en dataset limpio de productos
4,ventas_sin_producto,0,OK,Ventas sin producto asociado
5,totales_mal_calculados,0,OK,Diferencia entre total original y total recalc...


Reporte generado: c:\Users\EDELSON ORIHUELA\OneDrive\Documentos\PROYECTOS CV\GPT\Mini Proyecto – Analisis de Datos con Python\reports\quality_report.csv


## 12. Analisis descriptivo

En esta fase se calculan metricas generales y rankings principales para entender el comportamiento comercial de la ferreteria despues de la limpieza de datos.


In [5]:
from pathlib import Path

import pandas as pd
import numpy as np

PROJECT_ROOT = Path.cwd().parent if Path.cwd().name == "notebooks" else Path.cwd()
DATA_PROCESSED = PROJECT_ROOT / "data" / "processed"

ventas_limpias = pd.read_csv(DATA_PROCESSED / "ventas_limpias.csv")
productos_limpios = pd.read_csv(DATA_PROCESSED / "productos_limpios.csv")

ventas_limpias["fecha"] = pd.to_datetime(ventas_limpias["fecha"])

print("Ventas limpias:", ventas_limpias.shape)
print("Productos limpios:", productos_limpios.shape)


Ventas limpias: (1800, 18)
Productos limpios: (72, 6)


### 12.1 KPIs generales

Se calculan indicadores principales de ventas, unidades, transacciones, ticket promedio y margen estimado.


In [7]:
ventas_totales = ventas_limpias["total"].sum()
unidades_vendidas = ventas_limpias["cantidad"].sum()
transacciones = len(ventas_limpias)
promedio_venta = ventas_limpias["total"].mean()
precio_promedio_ponderado = ventas_totales / unidades_vendidas
margen_estimado_total = ventas_limpias["margen_estimado"].sum()
margen_estimado_pct = margen_estimado_total / ventas_totales

kpis_generales = pd.DataFrame(
    {
        "KPI": [
            "Ventas totales",
            "Unidades vendidas",
            "Transacciones",
            "Promedio de venta por linea",
            "Precio promedio ponderado",
            "Margen estimado",
            "Margen estimado %",
        ],
        "Valor": [
            ventas_totales,
            unidades_vendidas,
            transacciones,
            promedio_venta,
            precio_promedio_ponderado,
            margen_estimado_total,
            margen_estimado_pct,
        ],
    }
)

kpis_generales["Valor_formateado"] = [
    f"S/ {ventas_totales:,.2f}",
    f"{unidades_vendidas:,.0f}",
    f"{transacciones:,.0f}",
    f"S/ {promedio_venta:,.2f}",
    f"S/ {precio_promedio_ponderado:,.2f}",
    f"S/ {margen_estimado_total:,.2f}",
    f"{margen_estimado_pct:.2%}",
]

kpis_generales


,KPI,Valor,Valor_formateado
0,Ventas totales,128182.960000,"S/ 128,182.96"
1,Unidades vendidas,8056.000000,"8,056"
2,Transacciones,1800.000000,"1,800"
3,Promedio de venta por linea,71.212756,S/ 71.21
4,Precio promedio ponderado,15.911490,S/ 15.91
5,Margen estimado,35775.120000,"S/ 35,775.12"
6,Margen estimado %,0.279094,27.91%


**Lectura inicial:** El dataset registra S/ 128,182.96 en ventas, 8,056 unidades vendidas y 1,800 lineas de venta. El margen estimado total es de S/ 35,775.12, equivalente a 27.91% de las ventas. Esta lectura confirma que los resultados calculados en Python son consistentes con los KPIs del Proyecto 1 en Power BI.


### 12.2 Productos mas vendidos

Se identifican los productos con mayor venta total y mayor cantidad de unidades vendidas. Esto permite diferenciar productos que generan mas ingresos de productos que tienen mayor rotacion.


In [8]:
ranking_productos = (
    ventas_limpias
    .groupby(["id_producto", "producto", "categoria"], as_index=False)
    .agg(
        ventas_totales=("total", "sum"),
        unidades_vendidas=("cantidad", "sum"),
        transacciones=("producto", "count"),
        margen_estimado=("margen_estimado", "sum"),
        precio_promedio=("precio_unitario", "mean"),
    )
)

ranking_productos["margen_pct"] = (
    ranking_productos["margen_estimado"] / ranking_productos["ventas_totales"]
).round(4)

ranking_productos = ranking_productos.sort_values("ventas_totales", ascending=False)

ranking_productos.head(10)


,id_producto,producto,categoria,ventas_totales,unidades_vendidas,transacciones,margen_estimado,precio_promedio,margen_pct
15,P017,Cemento bolsa 42.5 kg,Materiales de construcción,20157.32,681,91,3813.32,29.576593,0.1892
17,P019,Pegamento cerámico 25 kg,Materiales de construcción,8293.77,353,45,1869.17,23.448222,0.2254
8,P009,Taladro percutor 650W,Herramientas eléctricas,5973.08,31,19,1478.08,192.270526,0.2475
9,P010,Amoladora angular 4.5 pulgadas,Herramientas eléctricas,5220.52,34,24,1208.52,153.826667,0.2315
16,P018,Arena fina saco 40 kg,Materiales de construcción,4261.74,447,55,864.54,9.561091,0.2029
0,P001,Martillo de acero 16 oz,Herramientas manuales,3414.96,126,59,1083.96,27.078814,0.3174
23,P025,Pintura látex blanca 1 gl,Pinturas,2874.21,59,27,868.21,48.745926,0.3021
18,P020,Yeso bolsa 20 kg,Materiales de construcción,2802.91,204,21,599.71,13.774762,0.2140
1,P002,Alicate universal 8 pulgadas,Herramientas manuales,2763.38,114,41,916.58,24.237073,0.3317
36,P038,Breaker termomagnético 20A,Electricidad,2732.56,82,16,764.56,33.342500,0.2798


In [9]:
ranking_productos_unidades = ranking_productos.sort_values(
    "unidades_vendidas", ascending=False
)

ranking_productos_unidades.head(10)


,id_producto,producto,categoria,ventas_totales,unidades_vendidas,transacciones,margen_estimado,precio_promedio,margen_pct
15,P017,Cemento bolsa 42.5 kg,Materiales de construcción,20157.32,681,91,3813.32,29.576593,0.1892
16,P018,Arena fina saco 40 kg,Materiales de construcción,4261.74,447,55,864.54,9.561091,0.2029
19,P021,Ladrillo pandereta unidad,Materiales de construcción,379.78,387,63,62.44,0.980000,0.1644
17,P019,Pegamento cerámico 25 kg,Materiales de construcción,8293.77,353,45,1869.17,23.448222,0.2254
34,P036,Cinta aislante negra,Electricidad,1166.02,326,56,481.42,3.570536,0.4129
47,P049,Tornillo drywall 1 pulgada ciento,Tornillería y fijaciones,2724.00,323,63,1173.60,8.424286,0.4308
31,P033,Cable THW 12 AWG metro,Electricidad,1173.35,323,64,398.15,3.593281,0.3393
35,P037,Foco LED 9W luz fría,Electricidad,2081.67,292,55,680.07,7.139818,0.3267
33,P035,Tomacorriente doble,Electricidad,2599.70,237,43,893.30,10.995814,0.3436
39,P041,Tubo PVC agua 1/2 pulgada,Gasfitería,1786.92,236,46,512.52,7.582826,0.2868


**Lectura:** El ranking por ventas muestra que Cemento bolsa 42.5 kg es el principal producto del negocio por ingresos. Sin embargo, al revisar unidades vendidas aparecen productos de menor precio con alta rotacion, como Arena fina saco 40 kg y Ladrillo pandereta unidad. Esta diferencia es importante porque un producto puede liderar en ingresos sin necesariamente ser el de mayor rotacion operativa.


### 12.3 Categorias con mayor venta

Se analiza el desempeno por categoria para identificar que lineas de negocio concentran mayores ventas, unidades y margen estimado.


In [10]:
resumen_categorias = (
    ventas_limpias
    .groupby("categoria", as_index=False)
    .agg(
        ventas_totales=("total", "sum"),
        unidades_vendidas=("cantidad", "sum"),
        transacciones=("producto", "count"),
        margen_estimado=("margen_estimado", "sum"),
        precio_promedio=("precio_unitario", "mean"),
    )
)

resumen_categorias["participacion_ventas_pct"] = (
    resumen_categorias["ventas_totales"] / resumen_categorias["ventas_totales"].sum()
).round(4)

resumen_categorias["margen_pct"] = (
    resumen_categorias["margen_estimado"] / resumen_categorias["ventas_totales"]
).round(4)

resumen_categorias = resumen_categorias.sort_values(
    "ventas_totales", ascending=False
)

resumen_categorias


,categoria,ventas_totales,unidades_vendidas,transacciones,margen_estimado,precio_promedio,participacion_ventas_pct,margen_pct
5,Materiales de construcción,41733.51,2512,327,8725.07,16.229083,0.3256,0.2091
2,Herramientas eléctricas,19839.05,99,67,4802.05,203.093284,0.1548,0.2421
3,Herramientas manuales,14020.88,667,255,4579.68,21.327961,0.1094,0.3266
0,Electricidad,13953.82,1571,297,4488.02,8.597441,0.1089,0.3216
6,Pinturas,10407.09,536,208,3380.39,19.079615,0.0812,0.3248
8,Tornillería y fijaciones,9188.04,1035,238,3767.64,8.791345,0.0717,0.4101
1,Gasfitería,7816.31,897,206,2422.91,8.583350,0.0610,0.3100
7,Seguridad industrial,6815.33,573,140,2222.03,12.443000,0.0532,0.3260
4,Jardinería,4408.93,166,62,1387.33,26.613065,0.0344,0.3147


**Lectura:** Materiales de construccion es la categoria con mayor venta total y concentra una parte importante del negocio. Herramientas electricas aparece como una categoria de alto ingreso pese a vender pocas unidades, lo que sugiere productos de mayor ticket promedio. En cambio, categorias como Electricidad o Tornilleria pueden tener mayor movimiento operativo por volumen de unidades.


### 12.4 Analisis mensual de ventas

Se agrupan las ventas por periodo mensual para observar evolucion en ventas, unidades, transacciones y margen estimado.


In [11]:
resumen_mensual = (
    ventas_limpias
    .groupby("periodo", as_index=False)
    .agg(
        ventas_totales=("total", "sum"),
        unidades_vendidas=("cantidad", "sum"),
        transacciones=("producto", "count"),
        margen_estimado=("margen_estimado", "sum"),
        ticket_promedio=("total", "mean"),
    )
)

resumen_mensual["variacion_ventas_pct"] = (
    resumen_mensual["ventas_totales"].pct_change()
).round(4)

resumen_mensual["media_movil_3m"] = (
    resumen_mensual["ventas_totales"].rolling(window=3, min_periods=1).mean()
).round(2)

resumen_mensual


,periodo,ventas_totales,unidades_vendidas,transacciones,margen_estimado,ticket_promedio,variacion_ventas_pct,media_movil_3m
0,2025-01,6208.13,464,94,1746.28,66.043936,NaN,6208.13
1,2025-02,8926.11,532,104,2396.65,85.827981,0.4378,7567.12
2,2025-03,7078.77,482,113,2002.56,62.643982,-0.2070,7404.34
3,2025-04,8216.60,441,104,2294.61,79.005769,0.1607,8073.83
4,2025-05,8301.45,489,99,2166.11,83.853030,0.0103,7865.61
5,2025-06,7033.02,480,99,2012.12,71.040606,-0.1528,7850.36
6,2025-07,9484.79,568,129,2685.51,73.525504,0.3486,8273.09
7,2025-08,7548.47,539,116,2112.30,65.073017,-0.2042,8022.09
8,2025-09,9501.31,563,134,2738.75,70.905299,0.2587,8844.86
9,2025-10,7640.06,474,125,2138.57,61.120480,-0.1959,8229.95


In [12]:
mejor_mes = resumen_mensual.loc[resumen_mensual["ventas_totales"].idxmax()]
peor_mes = resumen_mensual.loc[resumen_mensual["ventas_totales"].idxmin()]

print("Mejor mes:")
display(mejor_mes.to_frame().T)

print("Peor mes:")
display(peor_mes.to_frame().T)


Mejor mes:


,periodo,ventas_totales,unidades_vendidas,transacciones,margen_estimado,ticket_promedio,variacion_ventas_pct,media_movil_3m
10,2025-11,11146.95,621,138,2933.51,80.775,0.459,9429.44


Peor mes:


,periodo,ventas_totales,unidades_vendidas,transacciones,margen_estimado,ticket_promedio,variacion_ventas_pct,media_movil_3m
14,2026-03,5669.31,384,99,1584.84,57.265758,-0.268,6376.61


**Lectura:** El resumen mensual permite identificar meses de mayor y menor desempeno comercial. Esta informacion sera util para construir visualizaciones de tendencia y para comparar si los picos de ventas corresponden a mayor cantidad de unidades, mayor ticket promedio o una combinacion de ambos factores.


### 12.5 Productos con mayor margen estimado

Ademas de analizar ventas, se revisan los productos que mas aportan margen estimado. Esto ayuda a diferenciar productos que venden mucho de productos que contribuyen mas a la rentabilidad.


In [13]:
ranking_margen = ranking_productos.sort_values(
    "margen_estimado", ascending=False
)

ranking_margen.head(10)


,id_producto,producto,categoria,ventas_totales,unidades_vendidas,transacciones,margen_estimado,precio_promedio,margen_pct
15,P017,Cemento bolsa 42.5 kg,Materiales de construcción,20157.32,681,91,3813.32,29.576593,0.1892
17,P019,Pegamento cerámico 25 kg,Materiales de construcción,8293.77,353,45,1869.17,23.448222,0.2254
8,P009,Taladro percutor 650W,Herramientas eléctricas,5973.08,31,19,1478.08,192.270526,0.2475
9,P010,Amoladora angular 4.5 pulgadas,Herramientas eléctricas,5220.52,34,24,1208.52,153.826667,0.2315
47,P049,Tornillo drywall 1 pulgada ciento,Tornillería y fijaciones,2724.00,323,63,1173.60,8.424286,0.4308
0,P001,Martillo de acero 16 oz,Herramientas manuales,3414.96,126,59,1083.96,27.078814,0.3174
1,P002,Alicate universal 8 pulgadas,Herramientas manuales,2763.38,114,41,916.58,24.237073,0.3317
33,P035,Tomacorriente doble,Electricidad,2599.70,237,43,893.30,10.995814,0.3436
23,P025,Pintura látex blanca 1 gl,Pinturas,2874.21,59,27,868.21,48.745926,0.3021
16,P018,Arena fina saco 40 kg,Materiales de construcción,4261.74,447,55,864.54,9.561091,0.2029


In [14]:
comparativo_ventas_margen = ranking_productos[
    [
        "producto",
        "categoria",
        "ventas_totales",
        "unidades_vendidas",
        "margen_estimado",
        "margen_pct",
    ]
].head(10)

comparativo_ventas_margen


,producto,categoria,ventas_totales,unidades_vendidas,margen_estimado,margen_pct
15,Cemento bolsa 42.5 kg,Materiales de construcción,20157.32,681,3813.32,0.1892
17,Pegamento cerámico 25 kg,Materiales de construcción,8293.77,353,1869.17,0.2254
8,Taladro percutor 650W,Herramientas eléctricas,5973.08,31,1478.08,0.2475
9,Amoladora angular 4.5 pulgadas,Herramientas eléctricas,5220.52,34,1208.52,0.2315
16,Arena fina saco 40 kg,Materiales de construcción,4261.74,447,864.54,0.2029
0,Martillo de acero 16 oz,Herramientas manuales,3414.96,126,1083.96,0.3174
23,Pintura látex blanca 1 gl,Pinturas,2874.21,59,868.21,0.3021
18,Yeso bolsa 20 kg,Materiales de construcción,2802.91,204,599.71,0.2140
1,Alicate universal 8 pulgadas,Herramientas manuales,2763.38,114,916.58,0.3317
36,Breaker termomagnético 20A,Electricidad,2732.56,82,764.56,0.2798


**Lectura:** Los productos lideres por ventas tambien tienden a concentrar margen estimado, pero no siempre en la misma proporcion. Este analisis permite priorizar productos no solo por ingresos, sino tambien por contribucion economica. Para una ferreteria, esta mirada es clave al decidir abastecimiento, promociones o acciones comerciales.


### 12.6 Conclusiones del analisis descriptivo

- El negocio registra ventas totales superiores a S/ 128 mil durante el periodo analizado.
- Materiales de construccion es la categoria con mayor peso comercial.
- Cemento bolsa 42.5 kg lidera el ranking de ventas y tambien destaca por unidades vendidas.
- Existen productos de bajo precio con alta rotacion que son importantes para el movimiento operativo.
- Las herramientas electricas generan ventas relevantes con pocas unidades, lo que evidencia un ticket promedio mayor.
- El margen estimado representa aproximadamente 27.91% de las ventas totales.
- El analisis descriptivo permite pasar de indicadores generales a preguntas mas especificas sobre productos, categorias, temporalidad y rentabilidad.


### 12.7 Exportacion de tablas analiticas

Se guardan los principales resumenes generados para facilitar futuras visualizaciones, documentacion y revision externa.


In [15]:
ranking_productos.to_csv(DATA_PROCESSED / "ranking_productos.csv", index=False, encoding="utf-8-sig")
resumen_categorias.to_csv(DATA_PROCESSED / "resumen_categorias.csv", index=False, encoding="utf-8-sig")
resumen_mensual.to_csv(DATA_PROCESSED / "resumen_mensual.csv", index=False, encoding="utf-8-sig")
ranking_margen.to_csv(DATA_PROCESSED / "ranking_margen_productos.csv", index=False, encoding="utf-8-sig")

print("Tablas analiticas exportadas en:", DATA_PROCESSED)


Tablas analiticas exportadas en: c:\Users\EDELSON ORIHUELA\OneDrive\Documentos\PROYECTOS CV\GPT\Mini Proyecto – Analisis de Datos con Python\data\processed
